js/sw.js

```diff
@@ -204,49 +396,76 @@ const WORKER = `{
 
       self.postMessage({ state: currentState, validity });
     }
+    
+    if(e.data.type === "readContractState") {
+      const { state, key, contractId, validity, returnValidity } = e.data;
+      let stateValidity = { state, validity };
+      await globalThis.SmartWeave.readContractCalls[key](returnValidity ? stateValidity : state);
+      delete globalThis.SmartWeave.readContractCalls[key];
+    }
   });
 }`;
 
 export class Runtime {
   #state;
   #module;
 
-  constructor(source, state = {}, info = {}) {
+  constructor(source, state = {}, info = {}, executor, gateway) {
     this.#state = state;
     const sources = [WORKER, source];
     const blob = new Blob(sources, { type: "application/javascript" });
     this.#module = new Worker(
       URL.createObjectURL(blob),
       { eval: true, type: "module" },
     );
+    this.executor = executor;
+    this.gateway = gateway;
   }
 
   async resolveState() {
     this.#state = await new Promise((resolve) => {
-      this.#module.onmessage = (e) => {
-        resolve(e.data);
-        this.#module.terminate();
+      this.#module.onmessage = async (e) => {
+        if(e.data.readContractState) {
+          const { contractId, key, returnValidity, height, currentHeight } = e.data;
+          const { state, validity } = await this.executor.executeContract(contractId, height || currentHeight, true, this.gateway, returnValidity);
```

Comment:
If we pass a `false` value for the `clearCache` in the executeContract, it gets in an infinite loop. This is vital to take a look at why it does that when it has cached content before we merge this @littledivy